proof/refine/ARM/Ipc_R.thy

```diff
@@ -3156,6 +3156,461 @@ lemma maybeReturnSc_corres:
   apply (clarsimp simp: obj_at'_def)
   done
 
+lemma schedContextDonate_corres:
+  "corres dc \<top> \<top>
+   (sched_context_donate scp thread)
+   (schedContextDonate scp thread)"
+  unfolding sched_context_donate_def schedContextDonate_def
+  apply (clarsimp simp: get_sc_obj_ref_def set_tcb_obj_ref_thread_set)
+  apply (rule corres_guard_imp)
+    apply (rule corres_stateAssert_assume)
+     apply (rule corres_split [OF _ get_sc_corres])
+        apply (rule corres_split [OF _ corres_when2])
+           apply (rule corres_split
+                      [OF threadset_corresT
+                          update_sc_no_reply_stack_update_ko_at'_corres
+                            [where f'="scTCB_update (\<lambda>_. Some thread)"]])
+         apply (clarsimp simp: tcb_relation_def)
+         apply (clarsimp simp: tcb_cap_cases_def tcb_cte_cases_def)
+         apply (clarsimp simp: tcb_cap_cases_def tcb_cte_cases_def)
+         apply (clarsimp simp: sc_relation_def)
+         apply (clarsimp simp: )
+         apply (clarsimp simp: objBits_def objBitsKO_def)
+         apply (clarsimp simp: )
+   apply wpsimp
+   apply wpsimp
+         apply (clarsimp simp: sc_relation_def)
+    apply (rule corres_assert_opt_assume_l)
+  sorry (* awaiting spec change confirmation *)
+
+lemma replyNext_update_corres_Head:
+  "corres dc (reply_at rptr) (reply_at' rptr)
+   (set_reply_obj_ref reply_sc_update rptr (Some scptr))
+   (do reply <- getReply rptr;
+       setReply rptr (replyNext_update (\<lambda>_. Some (Head scptr)) reply)
+    od)"
+    unfolding update_sk_obj_ref_def
+  apply (clarsimp simp: )
+  apply (rule corres_guard_imp)
+  apply (rule corres_split [OF set_reply_corres get_reply_corres])
+  apply (clarsimp simp: reply_relation_def)
+  apply wpsimp+
+  apply (clarsimp simp: obj_at'_def replyPrev_same_def)
+  done
+
+lemma replyNext_update_corres_Next:
+  "corres dc (reply_at rptr) (reply_at' rptr)
+   (set_reply_obj_ref reply_sc_update rptr None)
+   (do reply <- getReply rptr;
+       setReply rptr (replyNext_update (\<lambda>_. Some (Next scptr)) reply)
+    od)"
+    unfolding update_sk_obj_ref_def
+  apply (clarsimp simp: )
+  apply (rule corres_guard_imp)
+  apply (rule corres_split [OF set_reply_corres get_reply_corres])
+  apply (clarsimp simp: reply_relation_def)
+  apply wpsimp+
+  apply (clarsimp simp: obj_at'_def replyPrev_same_def)
+  done
+
+lemma setSchedContext_sc_replies_relation_corres:
```

Comment:
Should it just be sc_relation in the name of this lemma?

Reply:
This is basically the same as one of the lemmas in my misc lemma PR (or an instantiation of it) and the name comes from there, I think...

Reply:
This is actually stronger because the preconditions are weaker (we don't need to show that R' holds for all sc's, only for the sc and sc' that are being modified). But otherwise this is exactly Miki's existing lemma.